core/controllers/reader.py

```diff
@@ -1371,6 +1371,130 @@ def get(self):
+        'GET': {}
+    }
+
+    @acl_decorators.can_play_exploration
+    def get(self, unique_progress_url_id):
+        """Handles GET requests. Fetches the logged-out learner's progress."""
+
+        logged_out_user_data = (
+            exp_fetchers.get_logged_out_user_progress(unique_progress_url_id))
+
+        if logged_out_user_data is None:
+            raise self.PageNotFoundException()
+
+        exploration = exp_fetchers.get_exploration_by_id(
+            logged_out_user_data.exploration_id, strict=False)
+
+        if logged_out_user_data is not None:
+            synced_exp_user_data = None
+            # If the latest exploration version is ahead of the most recently
+            # interacted exploration version.
+            if (
+                logged_out_user_data.most_recently_reached_checkpoint_exp_version is not None and # pylint: disable=line-too-long
+                logged_out_user_data.most_recently_reached_checkpoint_exp_version < exploration.version # pylint: disable=line-too-long
+            ):
+                synced_exp_user_data = (
+                    exp_services.sync_learner_checkpoint_progress_with_current_exp_version( # pylint: disable=line-too-long
+                        logged_out_user_data.exploration_id, unique_progress_url_id)) # pylint: disable=line-too-long
+            else:
+                synced_exp_user_data = logged_out_user_data
+
+            furthest_reached_checkpoint_exp_version = (
+                synced_exp_user_data.furthest_reached_checkpoint_exp_version)
+            furthest_reached_checkpoint_state_name = (
+                synced_exp_user_data.furthest_reached_checkpoint_state_name)
+            most_recently_reached_checkpoint_exp_version = (
+                synced_exp_user_data
+                    .most_recently_reached_checkpoint_exp_version)
+            most_recently_reached_checkpoint_state_name = (
+                synced_exp_user_data
+                    .most_recently_reached_checkpoint_state_name)
+
+        self.values.update({
+            'exploration_id': (
+                logged_out_user_data.exploration_id),
+            'furthest_reached_checkpoint_exp_version': (
+                furthest_reached_checkpoint_exp_version),
+            'furthest_reached_checkpoint_state_name': (
+                furthest_reached_checkpoint_state_name),
+            'most_recently_reached_checkpoint_exp_version': (
+                most_recently_reached_checkpoint_exp_version),
+            'most_recently_reached_checkpoint_state_name': (
+                most_recently_reached_checkpoint_state_name),
+        })
+
+        self.render_json(self.values)
+
+
+class SaveProgressEventHandler(base.BaseHandler):
+    """Responsible for storing progress of a logged-out learner."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+                'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        },
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
```

Comment:
There is only 'PUT' request here where you are assuming that the 'unique_progress_url_id' is already generated, I don't see a 'POST' request to create the Transient checkpoint model and generate the 'unique_progress_url_id' when the progress is saved for the first time.

Reply:
The unique_progress_url_id has to be generated in the frontend, and then sent as part of the payload when the Save Progress button is clicked. If the model doesn't already exist, a new model will be created using the exploration_id and the unique_progress_url_id. Thus, I believe we don't actually need a POST request here.

Reply:
@Manan-Rathi I think we decided to generate 'unique_progress_url_id' in the backend, it should be there in the comments thread of the TDD. To see how to implement this, you can take reference from how we generate a new exploration Id.

Reply:
Done. PTAL @paprajapati9